cmd/opampsupervisor/e2e_test.go

```diff
@@ -0,0 +1,319 @@
+			// so just check that it includes some strings we know to be unique to the remote config.
+			return strings.Contains(cfg, input.Name()) && strings.Contains(cfg, output.Name())
+		}
+
+		return false
+	}, 5*time.Second, 500*time.Millisecond, "Collector was not started with remote config")
+
+	n, err := input.WriteString("{\"body\":\"hello, world\"}\n")
+	require.NotZero(t, n, "Could not write to input file")
+	require.NoError(t, err)
+
+	require.Eventually(t, func() bool {
+		logRecord := make([]byte, 1024)
+		n, _ := output.Read(logRecord)
+
+		return n != 0
+	}, 10*time.Second, 500*time.Millisecond, "Log never appeared in output")
+}
+
+func TestSupervisorRestartsCollectorAfterBadConfig(t *testing.T) {
+	var healthReport atomic.Value
+	addr, connected, send := newServer(
+		t,
+		defaultConnectingHandler,
+		server.ConnectionCallbacksStruct{
+			OnMessageFunc: func(_ types.Connection, message *protobufs.AgentToServer) *protobufs.ServerToAgent {
+				if message.Health != nil {
+					healthReport.Store(message.Health)
+				}
+
+				return &protobufs.ServerToAgent{}
+			},
+		})
+
+	s := newSupervisor(t, "basic", map[string]string{"url": addr})
+	defer s.Shutdown()
+
+	waitForConnection(connected, true)
+
+	require.Eventually(t, func() bool {
+		health := healthReport.Load().(*protobufs.AgentHealth)
+
+		if health != nil {
+			return !health.Healthy && health.LastError != ""
+		}
+
+		return false
+	}, 5*time.Second, 250*time.Millisecond, "Supervisor never reported that the Collector was unhealthy")
+
+	cfg, hash, _, _ := createSimplePipelineConfig(t)
+
+	send(&protobufs.ServerToAgent{
+		RemoteConfig: &protobufs.AgentRemoteConfig{
+			Config: &protobufs.AgentConfigMap{
+				ConfigMap: map[string]*protobufs.AgentConfigFile{
+					"": {Body: cfg.Bytes()},
+				},
+			},
+			ConfigHash: hash,
+		},
+	})
+
+	require.Eventually(t, func() bool {
+		health := healthReport.Load().(*protobufs.AgentHealth)
+
+		if health != nil {
+			return health.Healthy && health.LastError == ""
+		}
+
+		return false
+	}, 5*time.Second, 250*time.Millisecond, "Supervisor never reported that the Collector became healthy")
+}
+
+func TestSupervisorConfiguresCapabilities(t *testing.T) {
+	var capabilities atomic.Uint64
+	addr, connected, _ := newServer(
+		t,
+		defaultConnectingHandler,
+		server.ConnectionCallbacksStruct{
+			OnMessageFunc: func(_ types.Connection, message *protobufs.AgentToServer) *protobufs.ServerToAgent {
+				capabilities.Store(message.Capabilities)
+
+				return &protobufs.ServerToAgent{}
+			},
+		})
+
+	s := newSupervisor(t, "nocap", map[string]string{"url": addr})
+	defer s.Shutdown()
+
+	waitForConnection(connected, true)
+
+	require.Eventually(t, func() bool {
+		cap := capabilities.Load()
+
+		return cap == uint64(protobufs.AgentCapabilities_AgentCapabilities_ReportsStatus)
+	}, 5*time.Second, 250*time.Millisecond)
+}
+
+func createSimplePipelineConfig(t *testing.T) (*bytes.Buffer, []byte, *os.File, *os.File) {
```

Comment:
Does this function create a Collector config?
It looks like it also has some side effects, like creating temporary files, and it is not completely clear what the purpose of that is.
It would be great if the function was documented.

Reply:
I've added some documentation to describe what's going on here. The gist is that this config creates a logs pipeline that allows us to test that there's a live Collector instance.